src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -3159,21 +3170,35 @@ long adjustMethodFlags(long flags) {
         return flags;
     }
 
-    long adjustClassFlags(long flags) {
-        boolean previewClassFile = minorVersion == ClassFile.PREVIEW_MINOR_VERSION;
+    long adjustClassFlags(ClassSymbol c, long flags) {
         if ((flags & ACC_MODULE) != 0) {
             flags &= ~ACC_MODULE;
             flags |= MODULE;
         }
-        if ((flags & ACC_IDENTITY) != 0 || (majorVersion < V66.major && (flags & INTERFACE) == 0)) {
+        if (((flags & ACC_IDENTITY) != 0 && !isMigratedValueClass(flags)) || (majorVersion < V66.major && (flags & INTERFACE) == 0)) {
             flags |= IDENTITY_TYPE;
-        } else if ((flags & INTERFACE) == 0 && allowValueClasses && previewClassFile && majorVersion >= V66.major) {
+        } else if (needsValueFlag(c, flags)) {
```

Comment:
isn't this redundant? E.g. isn't VALUE_CLASS set as soon as you see the annotation?

Reply:
no we need this path for the cases when we load a `real` value class